src/Couchbase.Lite.Shared/Query/NQuery.cs

```diff
@@ -0,0 +1,155 @@
+        {
+            Database = database;
+            _n1qlQueryExpression = n1qlQueryExpression;
+        }
+        #endregion
+
+        #region Override Methods
+        public override unsafe IResultSet Execute()
+        {
+            if (Database == null) {
+                throw new InvalidOperationException(CouchbaseLiteErrorMessage.InvalidQueryDBNull);
+            }
+
+            var options = C4QueryOptions.Default;
+            var paramN1ql = _n1qlQueryExpression.FLEncode();
+
+            var e = (C4QueryEnumerator*)ThreadSafety.DoLockedBridge(err =>
+            {
+                if (_disposalWatchdog.IsDisposed) {
+                    return null;
+                }
+
+                if (_c4Query == null) {
+                    Check();
+                }
+
+                var localOpts = options;
+                return NativeRaw.c4query_run(_c4Query, &localOpts, FLSlice.Null, err);
+            });
+
+            paramN1ql.Dispose();
+
+            if (e == null) {
+                return new NullResultSet();
+            }
+
+            var retVal = new QueryResultSet(this, ThreadSafety, e, _columnNames);
+            _history.Add(retVal);
+            return retVal;
+        }
+
+        public override string Explain()
+        {
+            return _n1qlQueryExpression;
+        }
+
+        protected override unsafe void Dispose(bool finalizing)
+        {
+            if (!finalizing) {
+                Stop();
+                ThreadSafety.DoLocked(() =>
+                {
+                    foreach (var e in _history) {
+                        e.Release();
+                    }
+
+                    _history.Clear();
+                    Native.c4query_release(_c4Query);
+                    _c4Query = null;
+                    _disposalWatchdog.Dispose();
+                });
+            } else {
+                // Database is not valid inside finalizer, but thread safety
+                // is guaranteed
+                Native.c4query_release(_c4Query);
+                _c4Query = null;
+            }
+        }
+        #endregion
+
+        #region Private Methods
+        private unsafe void Check()
+        {
+            ThreadSafety.DoLockedBridge(err =>
+            {
+                if (_disposalWatchdog.IsDisposed) {
+                    return true;
+                }
+
+                var query = Native.c4query_new2(Database.c4db, C4QueryLanguage.N1QLQuery, _n1qlQueryExpression, null, err);
+                if (query == null) {
+                    return false;
+                }
+
+                if (_columnNames == null) {
+                    _columnNames = CreateColumnNames(query);
+                }
+
+                Native.c4query_release(_c4Query);
+                _c4Query = query;
+                return true;
+            });
+        }
+
+        private unsafe Dictionary<string, int> CreateColumnNames(C4Query* query)
```

Comment:
Is there anything different here vs XQuery?

Reply:
N1QL query doesn't have FromImpl
`
var fromImpl = FromImpl;
Debug.Assert(fromImpl != null, "CreateColumnNames reached without a FROM clause received");
`